Add support of ACF Multilingual

Screenfeed opened the pull request:
Fixes https://github.com/polylang/polylang-pro/issues/2482.
Fixes #21 

ACFML is WPML's [compat' plugin](https://wpml.org/documentation/related-projects/translate-sites-built-with-acf/) for ACF.

This PR brings compatibility with WPML's "Translations" selector (none / copy / copy once / translate) by changing ACFML's data format into PLL's format.
ACFML adds a `wpml_cf_preferences` key to the fields, with `0` (none), `1` (copy), `2` (translate), or `3` (copy once) as value.

The new class `ACFFields` acts like the other classes, like `Posts` etc, and allows to translate ACFML's data into PLL's.
Since the fields' data is stored in post's `post_content` column as serialized array, `SELECT` queries made with `wpdb` look for entries containing `"wpml_cf_preferences"`. At the end of each batch, a single `wpdb` query updates the data for the whole batch. This is why the batch size has been reduced by 100, so the query doesn't grow too much.
ACFML's data is not removed.

Note:
ACFML adds a `acfml_field_group_mode` data to field groups. Values can be `translation` (Same fields across languages), `localization` (Different fields across languages), or `advanced` ([Expert](https://wpml.org/documentation/related-projects/translate-sites-built-with-acf/expert-translation-option/)). This setting allows to set the "Translations" selector of each field automatically. For example with "Same fields across languages", a `number` field is set to "Copy", but it will be set to "Copy once" if the group is set to "Different fields across languages". See [this post](https://wpml.org/documentation/related-projects/translate-sites-built-with-acf/recommended-custom-fields-translation-preferences-for-acf-and-wpml/) for the details.

Screenfeed committed:
Introduced the class `ACFFields`

Hug0-Drelon commented on src/ACFFields.php:
```diff
@@ -0,0 +1,138 @@
+<?php
+/**
+ * PHP version 5.6
```
Is it true?

Screenfeed replied:
PHPCS would alert us otherwise (I kept the same version than other files).

Hug0-Drelon replied:
I know there is a sniff that enforce us to specify PHP version on *each* files (which I found completely useless). I'm just surprised this is not aligne with Polylang current requirements.

Maybe @Chouby kept it to ensure a wide range of backward compatibility to migrate old website...

Screenfeed replied:
That's what I thought so I kept it.

Chouby replied:
1. The sniff [was explicitly added to our standard](https://github.com/polylang/polylang-cs/blob/ce1c3527eb07db5b35999195d6dc5923be6df62f/Polylang/ruleset.xml#L26) for new projects and don't have to match the Polylang standard which has to deal with an older code base. I'm not fan either to add this version but don't have a strong argument against. If we want change that, the we must update our CS and remove the PHP version in a separate PR. 
2. The current plugin version was written several years ago (when we released Polylang 3.4) and supports WP 5.8 and Polylang 3.4 as minimum versions. Both support PHP 5.6. If we don't need to increase these numbers, then we can keep them as is. But of course if we need to raise at least one of them, I'm not opposed to.

Hug0-Drelon replied:
Thanks for the explanation!

Hug0-Drelon commented on src/ACFFields.php:
```diff
@@ -0,0 +1,138 @@
+			$wpdb->prepare(
+				"SELECT COUNT(*)
+				FROM {$wpdb->posts}
+				WHERE post_type = 'acf-field' AND post_content LIKE %s",
```
```suggestion
				WHERE post_type = 'acf-field' AND CONTAINS (post_content, %s)",
```
`CONTAINS` *can* be faster.

Screenfeed replied:
1. `CONTAINS` doesn't exist in mysql, but `INSTR` does the same thing.
2. From what I've read, `CONTAINS` and `INSTR` do a fuzzy match, while we want an exact match.
3. Note: `CONTAINS` can be faster, but only if the column has a full text index. `post_content` seems to have it but I'm not 100% sure, I'm not familiar with this concept.

With point 2, I would keep `LIKE`.

Hug0-Drelon replied:
I thought `CONTAINS` was available for MySQL. Maybe it's an alias for `INSTR` (need to be checked though).
But regarding fuzzy search, I don't think so. Especially if you don't need pattern (like here with `self::ACFML_KEY`).

Hug0-Drelon replied:
I'll let @Chouby choose what he prefers.

Screenfeed replied:
That's what I've read in several places. However, [the doc](https://learn.microsoft.com/en-us/sql/t-sql/queries/contains-transact-sql?view=sql-server-ver16) says it can be used for fuzzy or precise search.

Hug0-Drelon replied:
From your source: "Searches for precise or fuzzy (less precise) matches to single words and phrases". Meaning it can do both. So in order to perform fuzzy search you have to provide `contains_search_condition` according to the doc. Also, careful with Microsoft documentation because SQL Server and Azure SQL aren't exactly the same as MySQL (or Maria DB for instance).

Hug0-Drelon commented on src/ACFFields.php:
```diff
@@ -0,0 +1,138 @@
+	protected function handle() {
+		global $wpdb;
+
+		$batch_size = absint( WPML_TO_POLYLANG_QUERY_BATCH_SIZE / 100 ); // 50 by default, to limit the size of the UPDATE query.
```
No need to compute this each time, let's move it in a private property populated in the constructor for instance.

Screenfeed replied:
Good idea. Done in [6638a2a](https://github.com/polylang/wpml-to-polylang/pull/25/commits/6638a2a752399df6f2a9e7ffe41c1adab2dadd76).

Hug0-Drelon commented on src/ACFFields.php:
```diff
@@ -0,0 +1,138 @@
+				FROM {$wpdb->posts}
+				WHERE post_type = 'acf-field' AND post_content LIKE %s
+				ORDER BY ID ASC
+				LIMIT %d, %d",
```
```suggestion
				LIMIT %d OFFSET %d",
```
If some day WordPress is kind enough to support PostgreSQL...

Screenfeed replied:
I didn't know this notation and I find it easier to understand. Done in [4f701b6](https://github.com/polylang/wpml-to-polylang/pull/25/commits/4f701b635f2c14fac1ec15efad50e69d3ddce0a1).

Screenfeed committed:
Calculate the batch size only once

Screenfeed committed:
Use the notation `LIMIT %d OFFSET %d` for better compatibility

Chouby commented on src/ACFFields.php:
```diff
@@ -0,0 +1,153 @@
+				implode( "\n", $values ), // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+				implode( "\n", array_keys( $values ) ) // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+			)
+		);
```
Is it possible to us `prepare()` as done in https://github.com/polylang/polylang/blob/3.7-rc1/include/Model/Languages.php#L990-L999. I prefer to avoid PHPCS exclusions when possible.

Screenfeed replied:
Done in [def4252](https://github.com/polylang/wpml-to-polylang/pull/25/commits/def4252b400967446a54c605ed6b6bf68a115a91).

Chouby commented on src/ACFFields.php:
```diff
@@ -0,0 +1,153 @@
+		3 => 'copy_once', // WPML_COPY_ONCE_CUSTOM_FIELD.
+	];
+
+	/**
+	 * Batch size.
+	 *
+	 * @var int
+	 */
+	protected $batch_size;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 0.6
+	 */
+	public function __construct() {
+		$this->batch_size = absint( WPML_TO_POLYLANG_QUERY_BATCH_SIZE / 100 ); // 50 by default, to limit the size of the UPDATE query.
+	}
```
Is it useful to have a property since it's used in only one method?

Screenfeed replied:
We put this here so it is done only once.

Screenfeed replied:
Changed in [85c419b](https://github.com/polylang/wpml-to-polylang/pull/25/commits/85c419b376580f43f9485aedb95d803434f666d9).

Chouby commented on src/ACFFields.php:
```diff
@@ -0,0 +1,153 @@
+	protected function handle() {
+		global $wpdb;
+
+		$offset  = absint( ( $this->step * $this->batch_size ) - $this->batch_size );
```
Isn't it necessary to override `getPercentage()` as you change the size?

Or maybe better: introduce a method `getBatchSize()` in `AbstractSteppable` overriden in this class?

Screenfeed replied:
HA! I totally missed it!

Screenfeed replied:
Done in [85c419b](https://github.com/polylang/wpml-to-polylang/pull/25/commits/85c419b376580f43f9485aedb95d803434f666d9).

Screenfeed committed:
Introduced `AbstractSteppable::getBatchSyze()`

Screenfeed committed:
Use `$wpdb->prepare()`
Notes:
1. The previous code add a bug, it should be `implode( ',', array_keys( $values ) )` instead of `implode( "\n", array_keys( $values ) )`.
2. Also, `addcslashes()` was useless: `esc_sql()` does it.

Screenfeed committed:
PHPCS

Commit:
Since 0.7

Screenfeed commented:
@wpsumo we weren't aware of this issue until now, so it hasn't been fixed.

wpsumo commented:
@Screenfeed It was mentioned in this which was considered closed and merged into this deployment https://github.com/polylang/wpml-to-polylang/issues/21

Options pages with ACF post object files was something for example that didnt work to migrate, thanks!

Screenfeed commented:
@wpsumo Since Polylang Pro doesn't support ACF's option pages yet, it wouldn't make sense to import their relationships for now. However, this is something we expect to do in the future, so I opened #34. No ETA, this is not something we expect to do soon though, we already have things in our tabs.
